### PR TITLE
Implemented Snapshot class & tests as well as functions & tests in the Game class that are related to Snapshot.

### PR DESCRIPTION
**Pull request by Anna-Bele**
This pull request implements functionalities and tests for the Snapshot class and the functionalities and tests in the Game class that are related to Snapshot. Tests should be passing for both Snapshot and Game classes.

### CODE REVIEW

**Review by noahklowden — DISMISSED** [review]
Overall looks good!

File: appGeo/src/app/snapshot/snapshot_class_declaration.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+   the state of the bubble at the time of a murder, specifically the players in the bubble.
+   assuming that dead players are not in bubbles(? verify), it would be a direct saving
+   of the bubble with no changes. */
+export class Snapshot {
+    snapshot_id: number;
+    snapshot_content: Bubble;
+
+    constructor(id: number, content: Bubble){
+        this.snapshot_id = id
+        this.snapshot_content = content
```
[the remark this answers is not in the document]
Comment (noahklowden) [reply appGeo/src/app/snapshot/snapshot_class_declaration.ts:13]: This is smart. I actually agree this—in general a few more testing would be due dilligence.

**Review by jasulcaf — APPROVED** [review]

File: appGeo/src/tests/snapshot.spec.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      );
+
+    assert.true(player_map.has(13), "snapshot should have player in bubble with id 13");
+
+    //snapshot should save a copy of the bubble at the time of murder, i.e. it should not update
+    assert.equal(bub.addPlayer(Mark), 2, "player should be added to original bubble");
+    assert.true(bub.removePlayer(Jack), "player should be removed from original bubble");
+
+    assert.true(player_map.has(13), "original player should still be in snapshot");
+    assert.false(player_map.has(25), "new player should not be in snapshot");
```
[the remark this answers is not in the document]
Comment (jasulcaf) [reply appGeo/src/tests/snapshot.spec.ts:42]: We could probably still explicitly define it as deep memory, just to be on safe side. But the test seems to have ran successfully.

[the remark this answers is not in the document]
Comment (noahklowden) [reply appGeo/src/tests/snapshot.spec.ts:42]: If this is the case, then I support the merge!

File: appGeo/src/app/snapshot/snapshot_class_declaration.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Anna-Bele) [reply appGeo/src/app/snapshot/snapshot_class_declaration.ts]: I just tested more robustly and realized the previous approach might not work due to possible list shallow copy. I made some changes by implementing a list deep copy. You can see the new changes in the newest commit.

File: appGeo/src/tests/snapshot.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Anna-Bele) [reply appGeo/src/tests/snapshot.spec.ts]: Yep, I changed my code, so now it implements deep copy. I realized my previous tests were not robust enough to catch this problem.

**Pull request merged**